src/execution/window_segment_tree.cpp

```diff
@@ -6,6 +6,178 @@
+    : WindowAggregateState(aggregate, bind_info, result_type), partition(0), row(0) {
+
+	// Locate the partition boundaries
+	idx_t start = 0;
+	if (partition_mask.AllValid()) {
+		partition_offsets.emplace_back(0);
+	} else {
+		idx_t entry_idx;
+		idx_t shift;
+		while (start < count) {
+			partition_mask.GetEntryIndex(start, entry_idx, shift);
+
+			//	If start is aligned with the start of a block,
+			//	and the block is blank, then skip forward one block.
+			const auto block = partition_mask.GetValidityEntry(entry_idx);
+			if (partition_mask.NoneValid(block) && !shift) {
+				start += ValidityMask::BITS_PER_VALUE;
+				continue;
+			}
+
+			// Loop over the block
+			for (; shift < ValidityMask::BITS_PER_VALUE && start < count; ++shift, ++start) {
+				if (partition_mask.RowIsValid(block, shift)) {
+					partition_offsets.emplace_back(start);
+				}
+			}
+		}
+	}
+
+	//	Initialise the vector for caching the results
+	results = make_unique<Vector>(result_type, partition_offsets.size());
+	partition_offsets.emplace_back(count);
+
+	//	Start the first aggregate
+	AggregateInit();
+}
+
+void WindowConstantAggregate::Sink(DataChunk &payload_chunk, SelectionVector *filter_sel, idx_t filtered) {
+	const auto chunk_begin = row;
+	const auto chunk_end = chunk_begin + payload_chunk.size();
+
+	if (!inputs.ColumnCount() && payload_chunk.ColumnCount()) {
+		inputs.Initialize(Allocator::DefaultAllocator(), payload_chunk.GetTypes());
+	}
+
+	AggregateInputData aggr_input_data(bind_info, Allocator::DefaultAllocator());
+	idx_t begin = 0;
+	idx_t filter_idx = 0;
+	auto partition_end = partition_offsets[partition + 1];
+	while (row < chunk_end) {
+		if (row == partition_end) {
+			AggegateFinal(*results, partition++);
+			AggregateInit();
+			partition_end = partition_offsets[partition + 1];
+		}
+		partition_end = MinValue(partition_end, chunk_end);
+		auto end = partition_end - chunk_begin;
+
+		inputs.Reset();
+		if (begin) {
+			for (idx_t c = 0; c < payload_chunk.ColumnCount(); ++c) {
+				inputs.data[c].Slice(payload_chunk.data[c], begin, end);
+			}
+		} else {
+			inputs.Reference(payload_chunk);
+		}
+		inputs.SetCardinality(end - begin);
+
+		// Slice to any filtered rows
+		SelectionVector sel;
+		if (filter_sel) {
+			//	Find the first value in [begin, end)
+			for (; filter_idx < filtered; ++filter_idx) {
+				auto idx = filter_sel->get_index(filter_idx);
+				if (idx >= begin) {
+					break;
+				}
+			}
+			sel.Initialize(filter_sel->data());
+			idx_t nsel = 0;
+			for (; filter_idx < filtered; ++filter_idx, ++nsel) {
+				auto idx = filter_sel->get_index(filter_idx);
+				if (idx >= end) {
+					break;
+				}
+			}
+
+			if (nsel != inputs.size()) {
+				inputs.Slice(sel, nsel);
+			}
+		}
+
+		//	Aggregate the filtered rows into a single state
+		const auto count = inputs.size();
+		if (aggregate.simple_update) {
+			aggregate.simple_update(inputs.data.data(), aggr_input_data, inputs.ColumnCount(), state.data(), count);
```

Comment:
Could we add some extra tests for constant window aggregates? I wonder how well some of these edge cases are covered by our current set of tests. e.g.:

* Mix of constant/non-constant aggregate (`sum` and `lead` over same `OVER` clause)
* Constant aggregate with non-`simple_update`
* Constant aggregate with not-nicely-aligned partitions (e.g. partitions are 73, 75, 77, 79, 81, 83, ... rows)
* Constant aggregate with large partitions (> vector size), also not nicely-aligned

Reply:
The benchmark has 3000 element partitions, and the row count is low enough that the hash groups have multiple partitions. But I can certainly add the mixtures.